kyo-tag/shared/src/main/scala/kyo/tags.scala

```diff
@@ -2,110 +2,129 @@ package kyo
 
 import scala.annotation.switch
 import scala.annotation.tailrec
+import scala.collection.immutable
 import scala.quoted.*
 
-case class Tag[T](tpe: String) extends AnyVal
+opaque type Tag[T] = String
 
 object Tag:
 
+    type Full[T] = Tag[T] | Set[T]
+
+    given [T, U]: CanEqual[Full[T], Full[U]] = CanEqual.derived
+
     import internal.*
 
     inline given apply[T]: Tag[T] = ${ tagImpl[T] }
 
-    extension [T](t1: Tag[T])
+    extension [T](t1: Full[T])
 
         def show: String =
-            val decoded = t1.tpe.drop(2).takeWhile(_ != ';')
-            fromCompact.getOrElse(decoded, decoded)
-
-        infix def <:<[U](t2: Tag[U]): Boolean =
-            t1 =:= t2 || isSubtype(t1, t2)
+            t1 match
+                case t1: String =>
+                    val decoded = t1.drop(2).takeWhile(_ != ';')
+                    fromCompact.getOrElse(decoded, decoded)
+                case t1: Set[T] =>
+                    t1.show
 
-        infix def =:=[U](t2: Tag[U]): Boolean =
-            (t1.tpe eq t2.tpe) || t1.tpe == t2.tpe
+        infix def <:<[U](t2: Full[U]): Boolean =
+            t1 match
+                case t1: String =>
+                    t2 match
+                        case t2: String => isSubtype(t1, t2)
+                        case t2: Set[?] => t2 >:> t1
+                case t1: Set[T] =>
+                    t1 <:< t2
 
-        infix def =!=[U](t2: Tag[U]): Boolean =
-            (t1.tpe ne t2.tpe) || t1.tpe != t2.tpe
+        infix def =:=[U](t2: Full[U]): Boolean =
+            (t1 eq t2) || t1 == t2
 
-        infix def >:>[U](t2: Tag[U]): Boolean =
-            t1 =:= t2 || isSubtype(t2, t1)
+        infix def =!=[U](t2: Full[U]): Boolean =
+            !(t1 =:= t2)
 
-    end extension
-
-    case class Union[T](tags: Seq[Tag[?]]) extends AnyVal
+        infix def >:>[U](t2: Full[U]): Boolean =
+            t2 <:< t1
 
-    object Union:
+        def erased: Tag[Any] = t1.asInstanceOf[Tag[Any]]
```

Comment:
Isn't this unsafe in the `Full[T]` case?

Reply:
Good catch! I've pushed a fix providing separate `erased` extensions for `Tag` and `Full`.